### DISCUSSION TITLE
公告错误：Matery theme的公告为空时依旧显示公告栏

### DISCUSSION THREAD

**Discussion by ferreroxc** [tangly1024/NotionNext#1962] [Q&A]
**描述bug**
简单说明bug的现象、相关的错误提示、日志等
- 我的Notion数据库中没有公告=>没有type为Notice的数据行
- 根据`/themes/matery/components/Announcement.js`，当notice为空时，应当返回<></>，也就是不显示公告栏
https://github.com/tangly1024/NotionNext/blob/2e7aa0b21e62cd2ae80a175f13ef7f29a191d0d4/themes/matery/components/Announcement.js#LL6C1-L10C4
```
const Announcement = ({ notice }) => {
  const { locale } = useGlobal()
  if (!notice) {
    return <></>
  }
```
- 但是页面依旧显示了公告栏

**复现步骤**
出现这个bug的操作步骤
- 创建一个不含type=Notice的Notion数据库
- 使用matery theme
- `yarn dev`，去`localhost:3000`查看网页
- 看到公告栏

**期望的正常结果**
希望按这个步骤，正常操作结果是什么
- 看不到公告栏

**截图**
相关的页面，应该的结果
<img width="260" alt="image" src="https://github.com/tangly1024/NotionNext/assets/32792279/741a760b-59e7-4dae-a76b-527fd8612efd">

<img width="803" alt="image" src="https://github.com/tangly1024/NotionNext/assets/32792279/9e782906-affb-4345-b99f-83184518e9ac">

**环境**

- 操作系统: [例如. iOS, Android, macOS, windows] Mac OS 13.4 (22F66)
- 浏览器 [例如. chrome, safari, firefox] chrome
- NotionNext版本 3.13.6

**补充说明**
与问题相关的其它说明

# 问题分析
## notice是{}
```
const Announcement = ({ notice }) => {
  const { locale } = useGlobal()
  if (!notice) {
    return <></>
  }
```
在这段代码里
- notice为{}，是一个empty object
- => !notice为false（判断empty object需要用其他方法）
- => if不成立，return了公告栏的div

但如果
- notice为null
- => !notice为true
- => if成立，return空，公告栏不显示
## 为什么notice是{}而不是null
在`lib/notion/getNotionData.js`中的`deepClone`把null变成了{}
https://github.com/tangly1024/NotionNext/blob/6357aa8cd1d544a5082b61d8cb5e16adfada75d6/lib/notion/getNotionData.js#L28
<img width="807" alt="image" src="https://github.com/tangly1024/NotionNext/assets/32792279/683fd864-52d9-4128-8fb7-d50daa8fe3f0">

https://github.com/tangly1024/NotionNext/blob/6357aa8cd1d544a5082b61d8cb5e16adfada75d6/lib/utils.js#L89
<img width="840" alt="image" src="https://github.com/tangly1024/NotionNext/assets/32792279/c4f96a2f-f0d1-4758-8021-592d21b86aac">

## 解决思路
1. 在`/themes/matery/components/Announcement.js`里面把`if(!notice)`改成判断object为空的方法
2. 在`deepClone`里面把`newObj = Array.isArray(obj) ? [] : {}`改成`newObj = Array.isArray(obj) ? [] : null`，但不知道这样会不会影响到其他代码逻辑

## 期待大家评论哪个解决方法更好，请 @tangly1024  把这个issue assign给我吧诶嘿

**Comment by tangly1024** [comment]
感谢 @ferreroxc 对此问题的认真讨论❤️

如您所言，两个方法都能解决问题。个人偏向第二种，没有公告就是`null`，而不返回一个`{}`对象。

代价是会影响代码逻辑的影响，这个影响也许不会很大，原因是绝大大部分的判断我都是用 `if (!object)` ，而非`Object.keys(object).length`等方法。
